### PR TITLE
feat(goal_planner): add bezier based pull over planner

### PR DESCRIPTION
**Pull request by soblin**
## Description

ros2 run autoware_behavior_path_goal_planner plot_map{1, 2} gives following figures.

image

image

## Related links

**Parent Issue:**

- Link

## Notes for reviewers

None.

## Interface changes

None.

## Effects on system behavior

None.

### CODE REVIEW

File: planning/sampling_based_planner/autoware_bezier_sampler/src/bezier_sampling.cpp [hunk 1]
```diff
@@ -58,6 +58,33 @@ std::vector<Bezier> sample(
...
+    (initial.pose.y() - final.pose.y()) * (initial.pose.y() - final.pose.y()));
+  // Tangent vectors
+  const Eigen::Vector2d initial_tangent_unit(std::cos(initial.heading), std::sin(initial.heading));
+  const Eigen::Vector2d final_tangent_unit(std::cos(final.heading), std::sin(final.heading));
+  // Unit vectors
+  const Eigen::Vector2d initial_normal_unit = initial_tangent_unit.unitOrthogonal();
+  const Eigen::Vector2d final_normal_unit = final_tangent_unit.unitOrthogonal();
+  double initial_tangent_length = initial_velocity * distance_initial_to_final;
+  double final_tangent_length = final_velocity * distance_initial_to_final;
+  double acceleration_length = acceleration * distance_initial_to_final;
```
Comment (maxime-clem) [inline planning/sampling_based_planner/autoware_bezier_sampler/src/bezier_sampling.cpp:76]: Everything can be made `const` here.
```suggestion
  const double distance_initial_to_final = std::sqrt(
    (initial.pose.x() - final.pose.x()) * (initial.pose.x() - final.pose.x()) +
    (initial.pose.y() - final.pose.y()) * (initial.pose.y() - final.pose.y()));
  // Tangent vectors
  const Eigen::Vector2d initial_tangent_unit(std::cos(initial.heading), std::sin(initial.heading));
  const Eigen::Vector2d final_tangent_unit(std::cos(final.heading), std::sin(final.heading));
  // Unit vectors
  const Eigen::Vector2d initial_normal_unit = initial_tangent_unit.unitOrthogonal();
  const Eigen::Vector2d final_normal_unit = final_tangent_unit.unitOrthogonal();
  const double initial_tangent_length = initial_velocity * distance_initial_to_final;
  const double final_tangent_length = final_velocity * distance_initial_to_final;
  const double acceleration_length = acceleration * distance_initial_to_final;
```

**Review by maxime-clem — APPROVED** [review]
Change to the `autoware_bezier_sampler` look good.

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/pull_over_planner/bezier_pull_over.cpp [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+      }
+      // add shoulder lane_id if not found
+      for (const auto & lane : shoulder_lanes) {
+        if (
+          std::find(point.lane_ids.begin(), point.lane_ids.end(), lane.id()) ==
+          point.lane_ids.end()) {
+          point.lane_ids.push_back(lane.id());
+        }
+      }
+    }
```
Comment (kosuke55) [inline planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/pull_over_planner/bezier_pull_over.cpp:282]: better not to set too many id.
you can copy the code from shift pull over.
refer: https://github.com/autowarefoundation/autoware.universe/pull/9360

Comment (soblin) [reply planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/pull_over_planner/bezier_pull_over.cpp]: fixed in 96f632d1b17f2702d987d14e1ff7be5b65be0842

**Review by kosuke55 — APPROVED** [review]

**Pull request merged**